pkg/controller/direct/dataform/repository_controller.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	// Get location
+	location := obj.Spec.Region
+
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		resourceID: resourceID,
+		projectID:  projectID,
+		gcpClient:  gcpClient,
+		location:   location,
+		desired:    obj,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, fmt.Errorf("AdapterForURL not supported for DataformRepository yet")
+}
+
+type Adapter struct {
+	resourceID string
+	projectID  string
+	location   string
+	gcpClient  *gcp.Client
+	desired    *krm.DataformRepository
+	actual     *dataformpb.Repository
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	req := &dataformpb.GetRepositoryRequest{Name: a.fullyQualifiedName()}
+	dataformrepositorypb, err := a.gcpClient.GetRepository(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting DataformRepository %q: %w", a.fullyQualifiedName(), err)
+	}
+
+	a.actual = dataformrepositorypb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating object", "u", u)
+
+	projectID := a.projectID
+	if projectID == "" {
+		return fmt.Errorf("project is empty")
+	}
+	if a.resourceID == "" {
+		return fmt.Errorf("resourceID is empty")
+	}
+
+	desired := a.desired.DeepCopy()
+	mapCtx := &direct.MapContext{}
+	resource := DataformRepositorySpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return fmt.Errorf("converting DataformRepository spec to api: %w", mapCtx.Err())
+	}
+
+	req := &dataformpb.CreateRepositoryRequest{
+		Parent:       a.getParent(),
+		Repository:   resource,
+		RepositoryId: a.resourceID,
+	}
+	_, err := a.gcpClient.CreateRepository(ctx, req)
+	if err != nil {
+		return fmt.Errorf("DataformRepository %s creation failed: %w", resource.Name, err)
+	}
+
+	status := &krm.DataformRepositoryStatus{}
+
+	// TODO(acpana): add observed state
+	// status.ObservedState.CreateTime = ToOpenAPIDateTime(updated.GetCreateTime())
+	// status.ObservedState.UpdateTime = ToOpenAPIDateTime(updated.GetUpdateTime())
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+
+	updateMask := &fieldmaskpb.FieldMask{}
+	// TODO(acpana): handle updates
+
+	desired := a.desired.DeepCopy()
+	mapCtx := &direct.MapContext{}
+	resource := DataformRepositorySpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return fmt.Errorf("converting DataformRepository spec to api: %w", mapCtx.Err())
+	}
+
+	req := &dataformpb.UpdateRepositoryRequest{UpdateMask: updateMask, Repository: resource}
+	_, err := a.gcpClient.UpdateRepository(ctx, req)
```

Comment:
Just to check - if this returns an operation, you should wait on it.

Reply:
It returns an object